csrc/sparse/cutlass/sparse_compressor.cu

```diff
@@ -0,0 +1,200 @@
+
+  int m = a.size(0);
+  int k = a.size(1);
+
+  using ProblemShape = Shape<int, int, int, int>;
+  using ElementA = ElementA_;
+  using LayoutTagA = cutlass::layout::RowMajor;
+
+  // Layouts for reference (non-sparse) tensors
+  using StrideA = cutlass::gemm::TagToStrideA_t<LayoutTagA>;
+  using StrideE = StrideA;
+
+  using Gemm = typename std::conditional<
+      std::is_same_v<ElementA, int8_t>,
+      typename sm90_int8_config_default<int8_t, cutlass::half_t,
+                                        ScaledEpilogue>::Cutlass3xGemm,
+      typename std::conditional<
+          std::is_same_v<ElementA, cutlass::float_e4m3_t>,
+          typename sm90_fp8_config_default<cutlass::float_e4m3_t,
+                                           cutlass::half_t,
+                                           ScaledEpilogue>::Cutlass3xGemm,
+          typename std::conditional<
+              std::is_same_v<ElementA, cutlass::half_t>,
+              typename sm90_fp16_config_default<cutlass::half_t,
+                                                cutlass::half_t,
+                                                ScaledEpilogue>::Cutlass3xGemm,
+              typename sm90_bf16_config_default<
+                  cutlass::bfloat16_t, cutlass::half_t,
+                  ScaledEpilogue>::Cutlass3xGemm>::type>::type>::type;
+
+  using ElementAB = typename Gemm::ElementAB;
+  using ElementD = typename Gemm::ElementD;
+
+  // Just a dummy value
+  int32_t n = 128;
+
+  int64_t lda = a.stride(0);
+
+  using StrideA = Stride<int64_t, Int<1>, int64_t>;
+  using StrideB = Stride<int64_t, Int<1>, int64_t>;
+  using StrideC = typename Gemm::StrideC;
+
+  StrideA a_stride{lda, Int<1>{}, 0};
+
+  using GemmKernel = typename Gemm::GemmKernel;
+  typename GemmKernel::ProblemShape prob_shape{m, n, k, 1};
+
+  using LayoutA = typename GemmKernel::CollectiveMainloop::LayoutA;
+  using LayoutE = typename GemmKernel::CollectiveMainloop::LayoutE;
+
+  using ElementE = typename GemmKernel::CollectiveMainloop::ElementE;
+  using SparseConfig = typename GemmKernel::CollectiveMainloop::SparseConfig;
+
+  LayoutA a_layout = SparseConfig::fill_layoutA(prob_shape);
+  LayoutE e_layout = SparseConfig::fill_layoutE(prob_shape);
+
+  // typename Gemm::GemmKernel::ProblemShape prob_shape{m, 1, k, 1};
+
+  // Offline compressor kernel
+  using CompressorUtility =
+      cutlass::transform::kernel::StructuredSparseCompressorUtility<
+          ProblemShape, ElementA, LayoutTagA, SparseConfig>;
+
+  using CompressorKernel =
+      cutlass::transform::kernel::StructuredSparseCompressor<
+          ProblemShape, ElementA, LayoutTagA, SparseConfig,
+          cutlass::arch::Sm90>;
+
+  using Compressor =
+      cutlass::transform::device::TransformUniversalAdapter<CompressorKernel>;
+
+  auto [M, N, K, L] = prob_shape;
+
+  StrideA stride_A;
+  StrideA stride_A_compressed;
+  StrideE stride_E;
+
+  stride_A =
+      cutlass::make_cute_packed_stride(StrideA{}, cute::make_shape(M, K, L));
```

Comment:
The best practice is to instantiate CuTe strides directly using the strides from the `torch::Tensor`. That way the strides will be correct even when using transposed tensors or when dealing with slices.

See how this is done in the dense w8a8 case:
https://github.com/vllm-project/vllm/blob/43b05fa314e90e551d87211e8bdde2e2bb5a0bdc/csrc/quantization/cutlass_w8a8/scaled_mm_c3x.cu#L125-L135

Reply:
This is the provided code by NVIDIA to compress a 2:4 tile and it's used for benchmarking the kernel. It works for all row-major matrices as by convention the first operand of sparse matmul in CUTLASS has to be. This is why I'm using similar implementation to the original provided code from cutlass examples.

Reply:
Ok you're right, I checked the actual GEMM kernel and you are doing it the right way there.